src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEnvironmentVariables()
+        {
+            var accessKeyId = Environment.GetEnvironmentVariable("AWS_ACCESS_KEY_ID");
+            var secretAccessKey = Environment.GetEnvironmentVariable("AWS_SECRET_ACCESS_KEY");
+            var sessionToken = Environment.GetEnvironmentVariable("AWS_SESSION_TOKEN");
+
+            if (accessKeyId == null && secretAccessKey == null && sessionToken == null)
+            {
+                return null;
+            }
+            if (secretAccessKey != null && accessKeyId == null)
+            {
+                throw new InvalidOperationException("When using MONGODB-AWS authentication if a secret access key is provided via environment variables then an access key ID must be provided also.");
+            }
+            if (accessKeyId != null && secretAccessKey == null)
+            {
+                throw new InvalidOperationException("When using MONGODB-AWS authentication if an access key ID is provided via environment variables then a secret access key must be provided also.");
+            }
+            if (sessionToken != null && (accessKeyId == null || secretAccessKey == null))
+            {
+                throw new InvalidOperationException("When using MONGODB-AWS authentication if a session token is provided via environment variables then an access key ID and a secret access key must be provided also.");
+            }
+
+            return new AwsCredentials(accessKeyId, SecureStringHelper.ToSecureString(secretAccessKey), sessionToken);
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEcsResponse()
+        {
+            var relativeUri = Environment.GetEnvironmentVariable("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI");
+            if (relativeUri == null)
+            {
+                return null;
+            }
+
+            var response = AwsHttpClientHelper.GetECSResponseAsync(relativeUri).GetAwaiter().GetResult();
+            var parsedResponse = BsonDocument.Parse(response);
+            var accessKeyId = parsedResponse.GetValue("AccessKeyId", null)?.AsString;
+            var secretAccessKey = parsedResponse.GetValue("SecretAccessKey", null)?.AsString;
+            var sessionToken = parsedResponse.GetValue("Token", null)?.AsString;
+
+            return new AwsCredentials(accessKeyId, SecureStringHelper.ToSecureString(secretAccessKey), sessionToken);
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEc2Response()
+        {
+            var response = AwsHttpClientHelper.GetEC2ResponseAsync().GetAwaiter().GetResult();
+            var parsedResponse = BsonDocument.Parse(response);
+            var accessKeyId = parsedResponse.GetValue("AccessKeyId", null)?.AsString;
+            var secretAccessKey = parsedResponse.GetValue("SecretAccessKey", null)?.AsString;
+            var sessionToken = parsedResponse.GetValue("Token", null)?.AsString;
```

Comment:
Idk whether we need or don't need making the same validation with the HTTP response as we do in CreateAwsCredentialsFromEnvironmentVariables. I will rely on Mikalai here